vkcs/provider.go

```diff
@@ -0,0 +1,281 @@
+type configer interface {
+	LoadAndValidate() error
+	GetRegion() string
+	ComputeV2Client(region string) (*gophercloud.ServiceClient, error)
+	ImageV2Client(region string) (*gophercloud.ServiceClient, error)
+	NetworkingV2Client(region string, sdn string) (*gophercloud.ServiceClient, error)
+	BlockStorageV3Client(region string) (*gophercloud.ServiceClient, error)
+	KeyManagerV1Client(region string) (*gophercloud.ServiceClient, error)
+	GetMutex() *mutexkv.MutexKV
+}
+
+// config uses openstackbase.Config as the base/foundation of this provider's
+type config struct {
+	auth.Config
+}
+
+var _ configer = &config{}
+
+// GetRegion is implementation of getRegion method
+func (c *config) GetRegion() string {
+	return c.Region
+}
+
+func (c *config) ComputeV2Client(region string) (*gophercloud.ServiceClient, error) {
+	return c.Config.ComputeV2Client(region)
+}
+
+func (c *config) ImageV2Client(region string) (*gophercloud.ServiceClient, error) {
+	return c.Config.ImageV2Client(region)
+}
+
+func (c *config) NetworkingV2Client(region string, sdn string) (*gophercloud.ServiceClient, error) {
+	client, err := c.Config.NetworkingV2Client(region)
+	if err != nil {
+		return client, err
+	}
+	client.MoreHeaders = map[string]string{
+		"X-SDN": sdn,
+	}
+	return client, err
+}
+
+func (c *config) BlockStorageV3Client(region string) (*gophercloud.ServiceClient, error) {
+	return c.Config.BlockStorageV3Client(region)
+}
+
+func (c *config) KeyManagerV1Client(region string) (*gophercloud.ServiceClient, error) {
+	return c.Config.KeyManagerV1Client(region)
+}
+
+func (c *config) GetMutex() *mutexkv.MutexKV {
+	return c.Config.MutexKV
+}
+
+func newConfig(d *schema.ResourceData, terraformVersion string) (configer, diag.Diagnostics) {
+	config := &config{
+		auth.Config{
+			CACertFile:       d.Get("cacert_file").(string),
+			ClientCertFile:   d.Get("cert").(string),
+			ClientKeyFile:    d.Get("key").(string),
+			Password:         d.Get("password").(string),
+			TenantID:         d.Get("project_id").(string),
+			Region:           d.Get("region").(string),
+			AllowReauth:      true,
+			MaxRetries:       maxRetriesCount,
+			TerraformVersion: terraformVersion,
+			SDKVersion:       meta.SDKVersionString(),
+			MutexKV:          mutexkv.NewMutexKV(),
+		},
+	}
+
+	if config.TenantID == "" {
+		config.TenantID = os.Getenv("OS_PROJECT_ID")
+	}
+	if config.UserDomainID == "" {
+		config.UserDomainID = os.Getenv("OS_USER_DOMAIN_ID")
+	}
+	if config.Password == "" {
+		config.Password = os.Getenv("OS_PASSWORD")
+	}
+	if config.Username == "" {
+		config.Username = os.Getenv("OS_USERNAME")
+	}
+	if config.Region == "" {
+		config.Region = os.Getenv("OS_REGION")
+	}
+
+	v, ok := d.GetOk("insecure")
+	if ok {
+		insecure := v.(bool)
+		config.Insecure = &insecure
+	}
+	v, ok = d.GetOk("auth_url")
+	if ok {
+		config.IdentityEndpoint = v.(string)
+	} else {
+		config.IdentityEndpoint = defaultIdentityEndpoint
+	}
```

Comment:
Why no to try take `IdentityEndpoint` from env `OS_AUTH_URL` before setting `defaultIdentityEndpoint` if provider's `auth_url` not specified?